androidenv: Implement cmdline-tools and patcher

hadilq opened the pull request:
- Replace cmdline-tools with tools because tools is obsolete now.
- Add patcher package
- Depend emulator package to systemImages
- Add emulator's executables

This is a part of https://github.com/NixOS/nixpkgs/issues/208047

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

hadilq committed:
androidenv: Implement cmdline-tools and patcher

- Replace cmdline-tools with tools because tools is obsolete now.
- Depend emulator package to systemImages

androidenv: fix issues on the PR

androidenv: reformat

androidenv: support excluding of `tools` package

androidenv: provide `tools`, and `build-tools`, dependencies

androidenv: replace includeTools with toolsVersion

androidenv: fix a typo

androidenv: add tests to check licenses and installed packages

androidenv: check if tests are running! this commit should fail!

androidenv: fix problems in the review https://github.com/NixOS/nixpkgs/pull/208793

androidenv: add test-suite to handle more tests around

androidenv: fix the test after couldn't running them with ofborg

Update pkgs/development/mobile/androidenv/build-tools.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

androidenv: Resolving https://github.com/NixOS/nixpkgs/pull/208793#discussion_r1065851539

Update pkgs/development/mobile/androidenv/cmdline-tools.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Update pkgs/development/mobile/androidenv/tools.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

androidenv: fix a typo

hadilq commented:
Hey @numinit , please review this one. Thanks. 

- I decided to merge this one first, then 394e8e0f619d7a3648835b8c52d2f2b46d329066 to show the complains of the `avdmanager` about `package.xml` files.
- Also `avdmanager` cannot create avd, but next PR will fix it.

In the end, I want to add some tests to run commands in the shell of `example/shell.nix`. I took a look on tests in `nixos/tests`, but couldn't find any of them that opens an example shell and run commands there! Could you guide me through the best choices to make the tests here?

numinit commented:
That's a good question! If you want to write a test for this (which would be much appreciated) you could create a checkPhase in the deploy-android-packages.nix.

numinit commented on pkgs/development/mobile/androidenv/cmdline-tools.nix:
```diff
@@ -0,0 +1,44 @@
+    export ANDROID_SDK_ROOT="$out/libexec/android-sdk"
+
+    # Wrap all scripts that require JAVA_HOME
+    for i in $ANDROID_SDK_ROOT/cmdline-tools/${cmdLineToolsVersion}/bin
```
`for i in $ANDROID_SDK_ROOT/cmdline-tools/${cmdLineToolsVersion}/bin; do`

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/cmdline-tools.nix:
```diff
@@ -0,0 +1,44 @@
+    # Wrap all scripts that require JAVA_HOME
+    for i in $ANDROID_SDK_ROOT/cmdline-tools/${cmdLineToolsVersion}/bin
+    do
+      find $i -maxdepth 1 -type f -executable | while read program
```
find $i -maxdepth 1 -type f -executable | while read program; do

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/cmdline-tools.nix:
```diff
@@ -0,0 +1,44 @@
+    do
+      find $i -maxdepth 1 -type f -executable | while read program
+      do
+        if grep -q "JAVA_HOME" $program
```
if grep -q "JAVA_HOME" $program; then

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/compose-android-packages.nix:
```diff
@@ -2,7 +2,7 @@
 }:
 
-{ toolsVersion ? "26.1.1"
+{ cmdLineToolsVersion ? "8.0"
```
Since the API for this is changing, can you update the nixpkgs manual? It uses the old toolsVersion.

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/compose-android-packages.nix:
```diff
@@ -293,27 +311,22 @@ rec {
       mkdir -p $out/bin
-      find $PWD/tools -not -path '*/\.*' -type f -executable -mindepth 1 -maxdepth 1 | while read i
+
+      for i in ${platform-tools}/bin/*
```
for i in ${platform-tools}/bin/*; do

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/compose-android-packages.nix:
```diff
@@ -293,27 +311,22 @@ rec {
       done
 
-      find $PWD/tools/bin -not -path '*/\.*' -type f -executable -mindepth 1 -maxdepth 1 | while read i
+      for i in ${emulator}/bin/*
```
for i in ${emulator}/bin/*; do

numinit commented on pkgs/development/mobile/androidenv/compose-android-packages.nix:
```diff
@@ -293,27 +311,22 @@ rec {
       done
 
-      for i in ${platform-tools}/bin/*
+      find $ANDROID_SDK_ROOT/cmdline-tools/${cmdLineToolsVersion}/bin -type f -executable | while read i
```
find $ANDROID_SDK_ROOT/cmdline-tools/${cmdLineToolsVersion}/bin -type f -executable | while read i; do

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/emulator.nix:
```diff
@@ -47,6 +47,16 @@ deployAndroidPackage {
+
+    mkdir -p $out/bin
+    cd $out/bin
+    find $out/libexec/android-sdk/emulator -type f -executable -mindepth 1 -maxdepth 1 | while read i
```
find $out/libexec/android-sdk/emulator -type f -executable -mindepth 1 -maxdepth 1 | while read i; do

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/tools.nix:
```diff
@@ -1,26 +1,42 @@
+    ''}
+
+    # Wrap all scripts that require JAVA_HOME
+    for i in bin
```
for i in bin; do

numinit commented on pkgs/development/mobile/androidenv/tools.nix:
```diff
@@ -1,26 +1,42 @@
+    # Wrap all scripts that require JAVA_HOME
+    for i in bin
+    do
+        find $i -maxdepth 1 -type f -executable | while read program
```
find $i -maxdepth 1 -type f -executable | while read program; do

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/tools.nix:
```diff
@@ -1,26 +1,42 @@
+    do
+        find $i -maxdepth 1 -type f -executable | while read program
+        do
+            if grep -q "JAVA_HOME" $program
```
if grep -q "JAVA_HOME" $program; then

hadilq replied:
Fixed on [41acb57](https://github.com/NixOS/nixpkgs/pull/208793/commits/41acb572e21fd2d1776a0e148d059a2188d68dad)

numinit commented on pkgs/development/mobile/androidenv/compose-android-packages.nix:
```diff
@@ -2,7 +2,7 @@
+{ cmdLineToolsVersion ? "8.0"
 , platformToolsVersion ? "33.0.3"
 , buildToolsVersions ? [ "33.0.1" ]
 , includeEmulator ? false
```
Can we still include a `toolsVersion ? null` after cmdLineToolsVersion and assert that it remains null? (And tell people to either remove it or switch to `cmdLinetoolsVersion` if it's provided?)

hadilq replied:
it's obsolete, and if you check the repo.json there's only one version of it available. However, if you download the `cmdline-tools` independently, then extract it, then try to download any packages with `sdkmanager`, then it always downloads and extracts the `tools`! That's why I didn't give any choice to people! Apparently it should always be there,even that it's obsolete. If you like I can check it twice.

numinit replied:
Oh, I'm not suggesting we still support it. I'm suggesting that we assert that people _don't_ try to provide it, and help them out if they do.

hadilq replied:
Try using a boolean instead of a nullable. check [952737f](https://github.com/NixOS/nixpkgs/pull/208793/commits/952737ff484a22eed6069771a05ffc3ae44a9236). Makes sense?

numinit replied:
I think you are misunderstanding. I also think we should fully drop `tools` as it is obsolete. But if a user provides it in their shell environment we want to tell them to upgrade using an assert failure.

numinit replied:
>it's obsolete, and if you check the repo.json there's only one version of it available. However, if you download the cmdline-tools independently, then extract it, then try to download any packages with sdkmanager, then it always downloads and extracts the tools! That's why I didn't give any choice to people! Apparently it should always be there,even that it's obsolete. If you like I can check it again.

Maybe I'm misunderstanding too? It always requires the tools?

hadilq replied:
These are the facts that can impact our decision here:
- `tools` is obsolete, but `sdkmanager` always download it in all SDKs.
- There are some old binaries in `tools` that are not available on any other packages, like `uiautomatorviewer`.
- There's only one version of `tools` available in `repo.json`.
- In the `repository2-1.xml`, which is the source for `repo.json`, you can find that `built-tools` are depending on `tools`. Notice `build-tools` is mandatory in the SDK. Maybe that's why `sdkmanager` keeps the `tools` in all SDKs.
```xml
  <remotePackage path="build-tools;30.0.3">
    <!--Generated from bid:6966805, branch:rvc-qpr1-release-->
    <type-details xsi:type="generic:genericDetailsType"/>
    <revision>
      <major>30</major>
      <minor>0</minor>
      <micro>3</micro>
    </revision>
    <display-name>Android SDK Build-Tools 30.0.3</display-name>
    <uses-license ref="android-sdk-license"/>
    <dependencies>
      <dependency path="tools"/>
    </dependencies>
```

so there's no choice for the version of `tools`, and that's why I choose a boolean instead of a nullable. Maybe it's a good idea in the future to read the `dependencies` and build up the packages tree, without asking people to fill out the dependencies by hand! Let me know what do you think.

hadilq replied:
Also I just noticed
```xml
  <remotePackage obsolete="true" path="tools">
    <!--Generated from:ab bid:4333796 branch:git_studio-sdk-release-->
    <type-details xsi:type="generic:genericDetailsType"/>
    <revision>
      <major>26</major>
      <minor>1</minor>
      <micro>1</micro>
    </revision>
    <display-name>Android SDK Tools</display-name>
    <uses-license ref="android-sdk-license"/>
    <dependencies>
      <dependency path="patcher;v4"/>
      <dependency path="emulator"/>
      <dependency path="platform-tools">
        <min-revision>
          <major>20</major>
        </min-revision>
```
which means `tools` is depending on the `patcher`, `emulator`, and `platform-tools`. Let me create a commit to make them visible to the `tools`.

hadilq replied:
I specified those dependencies in [9495595](https://github.com/NixOS/nixpkgs/pull/208793/commits/9495595145ec54bac03dfe08ced5ee8319ae41f3)

numinit replied:
>  Maybe it's a good idea in the future to read the dependencies and build up the packages tree, without asking people to fill out the dependencies by hand! Let me know what do you think.

This is a good idea!

numinit replied:
Alright, I see now. That's confusing. I would still specify toolsVersion in case there _is_ an update at some point then, and because it makes composeAndroidPackages compatible with existing code.

hadilq commented:
Build failed for some unknown reason! see https://gist.github.com/GrahamcOfBorg/66ee252b2a8e53188f0a8d749412a6c1?permalink_comment_id=4422732#gistcomment-4422732

hadilq commented:
@numinit I prepared a test that's related to this PR, because before this PR `sdkmanager` is raising `Exception in thread "main" java.lang.NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema`. so here is the first place we can check that all the licenses are accepted. This is the diff

```diff
diff --git a/pkgs/development/mobile/androidenv/default.nix b/pkgs/development/mobile/androidenv/default.nix
index 90064becfb5..9eda9591095 100644
--- a/pkgs/development/mobile/androidenv/default.nix
+++ b/pkgs/development/mobile/androidenv/default.nix
@@ -19,4 +19,8 @@ rec {
     platformVersions = [ "28" ];
     abiVersions = [ "x86" "x86_64"];
   };
+
+  example-shell = pkgs.callPackage ./examples/shell.nix {};
+
+  tests = example-shell.tests;
 }
diff --git a/pkgs/development/mobile/androidenv/examples/shell.nix b/pkgs/development/mobile/androidenv/examples/shell.nix
index ab953eb620b..d818554b24f 100644
--- a/pkgs/development/mobile/androidenv/examples/shell.nix
+++ b/pkgs/development/mobile/androidenv/examples/shell.nix
@@ -138,11 +138,21 @@ pkgs.mkShell rec {

     # Write out local.properties for Android Studio.
     cat <<EOF > local.properties
-# This file was automatically generated by nix-shell.
-sdk.dir=$ANDROID_SDK_ROOT
-ndk.dir=$ANDROID_NDK_ROOT
-cmake.dir=$cmake_root
-EOF
+    # This file was automatically generated by nix-shell.
+    sdk.dir=$ANDROID_SDK_ROOT
+    ndk.dir=$ANDROID_NDK_ROOT
+    cmake.dir=$cmake_root
+    EOF
   '';
+
+  passthru.tests = {
+    sdkmanager-licenses-test = pkgs.runCommand "androidenv-full-sdk-shell-test" {
+      buildInputs = [ androidSdk platformTools jdk ];
+    } ''
+      [[ $(sdkmanager --licenses) =~ "All SDK package licenses accepted." ]] \
+        || (echo "At least one of SDK package licenses are not accepted." && exit 1)
+      touch $out
+    '';
+  };
 }

if you think it's okay to have it in this PR I can commit it.

numinit commented:
That sounds like a good idea.

hadilq commented:
I can run the tests by `nix-build -A androidenv.tests` command and it fails as expected after [6a1993e](https://github.com/NixOS/nixpkgs/pull/208793/commits/6a1993e7152bd97d48e65b3c7029b28905ce6f2a), but it's not failing here and when running `nix-build -A phoronix-test-suite.tests`! Investigating more!

hadilq commented:
@ofborg build androidenv.tests

numinit reviewed (approved):
Looks good, great work.

numinit commented on pkgs/development/mobile/androidenv/examples/shell.nix:
```diff
@@ -138,11 +138,157 @@ pkgs.mkShell rec {
+    sdkmanager-packages-test = pkgs.runCommand "sdkmanager-packages-test" {
+      buildInputs = [ androidSdk jdk ];
+    } ''
+      output=$(sdkmanager --list)
```
Double quotes around the $(...)

hadilq replied:
fixed on [a39b895](https://github.com/NixOS/nixpkgs/pull/208793/commits/a39b895f77668a0b58b53702729b9d5d7101963d)

numinit commented on pkgs/development/mobile/androidenv/examples/shell.nix:
```diff
@@ -138,11 +138,157 @@ pkgs.mkShell rec {
+    sdkmanager-licenses-test = pkgs.runCommand "sdkmanager-licenses-test" {
+      buildInputs = [ androidSdk jdk ];
+    } ''
+      if [[ ! $(sdkmanager --licenses) =~ "All SDK package licenses not accepted." ]]; then
```
Double quotes around $(...)

hadilq replied:
fixed on [a39b895](https://github.com/NixOS/nixpkgs/pull/208793/commits/a39b895f77668a0b58b53702729b9d5d7101963d)

numinit commented on pkgs/development/mobile/androidenv/examples/shell.nix:
```diff
@@ -138,11 +138,157 @@ pkgs.mkShell rec {
+      output=$(sdkmanager --list)
+      installed_packages_section=$(echo "''${output%%Available Packages*}" | awk 'NR>4 {print $1}')
+
+      if [[ ! $installed_packages_section =~ "build-tools;30.0.3" ]]; then
```
Can you do something like:

```
for section in 'ndk-bundle' 'platform-tools' 'platforms;android-23' ...; do
  if [[ ! "$installed_packages_section =~ "$section" ]; then
    echo "$section was not installed"
  fi
done
```

hadilq replied:
fixed on [a39b895](https://github.com/NixOS/nixpkgs/pull/208793/commits/a39b895f77668a0b58b53702729b9d5d7101963d)

hadilq commented:
@ofborg build androidenv.test-suite.passthru.tests

hadilq commented:
I couldn't make it work with ofborg! However, we have integration tests that can be ran locally now. So, it's ready @numinit, please approve it if it's okay. Thanks!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1684

SuperSandro2000 commented on pkgs/development/mobile/androidenv/build-tools.nix:
```diff
@@ -19,6 +19,9 @@ deployAndroidPackage {
     ''}
+
+    cd $out/libexec/android-sdk
+    ${postInstall}
   '';
```
```suggestion
  '' + postInstall;
```

SuperSandro2000 commented on pkgs/development/mobile/androidenv/cmdline-tools.nix:
```diff
@@ -0,0 +1,41 @@
+    export ANDROID_SDK_ROOT="$out/libexec/android-sdk"
+
+    # Wrap all scripts that require JAVA_HOME
+    for i in $ANDROID_SDK_ROOT/cmdline-tools/${cmdLineToolsVersion}/bin; do
+      find $i -maxdepth 1 -type f -executable | while read program; do
```
```suggestion
      find $ANDROID_SDK_ROOT/cmdline-tools/${cmdLineToolsVersion}/bin -maxdepth 1 -type f -executable | while read program; do
```
Would this do the same?

hadilq replied:
Eagle eyes! good job and thank you. Fixed [a19f22c](https://github.com/NixOS/nixpkgs/pull/208793/commits/a19f22c9c96f73e903099f8a825f500b907d6862)

SuperSandro2000 commented on pkgs/development/mobile/androidenv/cmdline-tools.nix:
```diff
@@ -0,0 +1,41 @@
+
+    # Wrap sdkmanager script
+    wrapProgram $ANDROID_SDK_ROOT/cmdline-tools/${cmdLineToolsVersion}/bin/sdkmanager \
+      --prefix PATH : ${pkgs.jdk11}/bin \
```
```suggestion
      --prefix PATH : ${lib.makeBinPath [ pkgs.jdk11 ]} \
```

SuperSandro2000 commented on pkgs/development/mobile/androidenv/tools.nix:
```diff
@@ -1,26 +1,42 @@
+    # Wrap monitor script
+    wrapProgram $PWD/monitor \
+      --prefix PATH : ${pkgs.jdk8}/bin \
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ pkgs.xorg.libX11 pkgs.xorg.libXtst ]}
```
```suggestion
      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath with pkgs; [ xorg.libX11 xorg.libXtst ]}
```

SuperSandro2000 commented on pkgs/development/mobile/androidenv/tools.nix:
```diff
@@ -1,26 +1,42 @@
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ pkgs.xorg.libX11 pkgs.xorg.libXtst ]}
+
+    # Patch all script shebangs
+    patchShebangs .
```
Can't we limit this to a subdirectory?

hadilq replied:
First of all I have to mention that `tools` is obsolete so I didn't change its code. I just copied the entire file from `tools/26.nix` to here. Second this is its structure:

```
dr-xr-xr-x 1 root root    216 Dec 31  1969 .
dr-xr-xr-x 1 root root     68 Dec 31  1969 ..
-r-xr-xr-x 1 root root   1313 Dec 31  1969 .monitor-wrapped
-r--r--r-- 1 root root 829319 Dec 31  1969 NOTICE.txt
-r-xr-xr-x 1 root root   4909 Dec 31  1969 android
dr-xr-xr-x 1 root root    420 Dec 31  1969 bin
-r-xr-xr-x 1 root root 629936 Dec 31  1969 emulator
-r-xr-xr-x 1 root root 414688 Dec 31  1969 emulator-check
dr-xr-xr-x 1 root root   3478 Dec 31  1969 lib
-r-xr-xr-x 1 root root  12191 Dec 31  1969 mksdcard
-r-xr-xr-x 1 root root   1159 Dec 31  1969 monitor
dr-xr-xr-x 1 root root    216 Dec 31  1969 proguard
-r--r--r-- 1 root root    138 Dec 31  1969 source.properties
dr-xr-xr-x 1 root root    222 Dec 31  1969 support
```
As you can see executables are all over the place so it looks like `patchShebangs .` is correct.

hadilq commented on pkgs/development/mobile/androidenv/tools.nix:
```diff
@@ -1,26 +1,42 @@
+  patchInstructions = ''
+    ${lib.optionalString (os == "linux") ''
+      # Auto patch all binaries
+      autoPatchelf .
```
The same as https://github.com/NixOS/nixpkgs/pull/208793#discussion_r1066435900

SuperSandro2000 commented:
The commit history is currently a bit of a mess. I could squash merge everything into one commit or it gets cleaned up and can be merged with a merge commit.

hadilq commented:
Hey @SuperSandro2000, I'm always okay to rebase and squash commits in a PR. I fact, I think it's a better approach than the freedom of keeping commits in a PR. Please do it.

Regarding the slow process of merging PRs, I have a little bit of time to share here, so I can own the androidenv directory to make the progress faster, if you guys are okay with that.

hadilq commented:
FYI, I just rebased, squashed and force pushed as suggested before, but two commits from other devs end up in my PR! Let me see if I can fix this mess! Sorry guys!

hadilq commented:
FYI, running pull rebase multiple time justified my purpose for git :D Let's see if we can land this PR.

numinit commented:
LGTM!
